bpf: added --bpf-lb-bypass-fib-lookup flag

skuffe opened the pull request:
This commit adds a boolean flag, which when enabled will cause the BPF nodeport reverse NAT handling to attempt to use a MAC address which was previously associated with the remote endpoint when sending its reply, instead of doing a FIB lookup.

If the flag is toggled off, this optimization is disabled, and a FIB lookup is always done.

This is useful if Cilium is to be used on nodes which default gateway is a FHRP VIP/MAC, such as with VRRP or Cisco's HSRP.

Fixes: #14911

Signed-off-by: Danni Skov Høglund <skuffe@pwnz.dk>

```release-note
Added --bpf-lb-bypass-fib-lookup flag, which toggles the BPF nodeport reverse NAT FIB lookup optimization
```

skuffe committed:
bpf: added --bpf-lb-bypass-fib-lookup flag

This commit adds a boolean flag, which when enabled will cause the BPF
nodeport reverse NAT handling to attempt to use a MAC address which was
previously associated with the remote endpoint when sending its reply,
instead of doing a FIB lookup.

If the flag is toggled off, this optimization is disabled, and a FIB
lookup is always done.

This is useful if Cilium is to be used on nodes which default gateway
is a FHRP VIP/MAC, such as with VRRP or Cisco's HSRP.

Fixes: #14911

Signed-off-by: Danni Skov Høglund <skuffe@pwnz.dk>

borkmann commented on bpf/lib/nodeport.h:
```diff
@@ -1649,7 +1654,8 @@ static __always_inline int rev_nodeport_lb4(struct __ctx_buff *ctx, int *ifindex
 
-		dmac = map_lookup_elem(&NODEPORT_NEIGH4, &ip4->daddr);
+		if (fib_lookup_bypass())
+			dmac = map_lookup_elem(&NODEPORT_NEIGH4, &ip4->daddr);
```
IPv6 counterpart is missing.

skuffe replied:
Missed that woops, added it in just now.

borkmann commented on pkg/defaults/defaults.go:
```diff
@@ -405,4 +405,7 @@ const (
 	KubeProxyReplacementHealthzBindAddr = ""
+
+	// EnableBPFBypassFIBLookup instructs Cilium to enable the FIB lookup bypass optimization for nodeport reverse NAT handling.
+	EnableBPFBypassFIBLookup = true
```
should be false by default, no?

Could you also update the Documentation/gettingstarted/kubeproxy-free.rst doc so that this knob's purpose is documented for users? Thx

skuffe replied:
If the default is "true" the behavior before these changes will be retained. IE. the optimization will be done by default. That way only the users with a FHRP gateway setup will need to set this to false.

I will take a look at the doc and get that updated!

skuffe replied:
@borkmann would you prefer to have any doc related commits in the same pull, and sqashed into the same single commit?

borkmann replied:
Ack, true, should have had my coffee first. :) Yeah, can be same commit or separate one, but within the same PR would be good.

skuffe committed:
docs & helm cm template + values.yaml

This commit adds a section to the Getting Started -> Kubernetes Without
kube-proxy, describing what this flag does and when to use it.

Also added the required bits to the Helm template and values.yaml

Signed-off-by: Danni Skov Høglund <skuffe@pwnz.dk>

brb reviewed (approved):
Thanks!

Have you tested both SNAT and DSR mode?

kkourt reviewed (approved):
Thanks!
